### PR TITLE
Allow sending configuration backups to email

### PR DESCRIPTION
**Pull request by machadovilaca**
screen shot 2018-08-08 at 11 35 49

### CODE REVIEW

**Comment by fichtner** [comment]
@machadovilaca we need to make this a plugin if you want to use gnupg, which should be easily possible with the backup framework

File: src/www/diag_backup.php [hunk 1]
```diff
@@ -521,6 +521,9 @@ function restore_config_section($section_name, $new_contents)
                         elseif ($field['type'] == 'password'):?>
 
                         <input name="<?=$fieldId;?>" type="password" value="<?=$field['value'];?>" />
+<?php
```
Comment (fichtner) [inline src/www/diag_backup.php]: if you send a separate pull request we can merge this directly

File: src/opnsense/mvc/app/library/OPNsense/Backup/PHPMailer/Exception.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?php
```
Comment (fichtner) [inline src/opnsense/mvc/app/library/OPNsense/Backup/PHPMailer/Exception.php]: phpmailer / phpmailer6 packages exists as well, it would be cleaner to use this instead of embedding a fixed version here. :)

File: src/www/diag_backup.php (hunk 1 above)
Comment (machadovilaca) [reply src/www/diag_backup.php]: #2616

File: src/opnsense/mvc/app/library/OPNsense/Backup/PHPMailer/Exception.php (hunk 2 above)
Comment (machadovilaca) [reply src/opnsense/mvc/app/library/OPNsense/Backup/PHPMailer/Exception.php]: ok, thanks, will change it

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            "type" => "text",
+            "label" => gettext("GPG Email"),
+            "value" => null
+        );
+        $fields[] = array(
+            "name" => "GpgPublicKey",
+            "type" => "textarea",
+            "label" => gettext("GPG Public Key"),
+            "value" => null
+        );
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: As an alternative, you should think about S/MIME since it does not require additional software in common mail clients.
OpenSSL / LibreSSL can do it for you if you don't want to do it manually: `man openssl-smime`

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                'libtasn1-4.13.txz', 'p11-kit-0.23.12.txz', 'libunistring-0.9.10.txz',
+                'libidn2-2.0.5.txz', 'libgpg-error-1.32.txz', 'libassuan-2.5.1.txz',
+                'tpm-emulator-0.7.4_2.txz', 'trousers-0.3.14_2.txz', 'pinentry-tty-1.1.0.txz',
+                'pinentry-1.1.0_1.txz', 'npth-1.6.txz', 'gnutls-3.5.18.txz', 'gnupg-2.2.9_1.txz'
+            );
+
+            foreach ($dependencies as $dependency) {
+                exec('pkg add ' . $link . $dependency . ' > /dev/null 2>&1');
+            }
+        }
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: dependencies should be handled via a package dependency. No installing in the code!

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        if ($cnf->isValid()) {
+            $config = $cnf->object();
+            if (isset($config->system->remotebackup) && isset($config->system->remotebackup->MailEnabled)
+                    && !empty($config->system->remotebackup->MailEnabled)) {
+                $confdata = file_get_contents('/conf/config.xml');
+                $return = self::sendEmail($config->system->remotebackup, $confdata);
+            }
+        }
+
+        return array($return);
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: you should not use a php keyword as a variable name.

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            $mail->Password = $smtpPassword;
+        }
+
+        $gpgPublicKeyFile = "key.asc";
+        if ($gpgPublicKey != "") {
+            file_put_contents($gpgPublicKeyFile, $gpgPublicKey);
+        }
+
+        exec('gpg2 --import < ' . $gpgPublicKeyFile);
+        exec('echo "' . $confdata . '" | gpg2 --trust-model always --batch --yes --output backup --encrypt --recipient ' . $gpgEmail);
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: looks like high danger zone - the problem here is that it looks like it is vulnerable to command injection.

Comment (machadovilaca) [reply src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: should I try to find another way to do this or input validation would be enough?

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php (hunk 5 above)
Comment (machadovilaca) [reply src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: changed

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php (hunk 4 above)
Comment (machadovilaca) [reply src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: I expect this to be fixed when making this a plugin as suggested in https://github.com/opnsense/core/pull/2615#issuecomment-411363094

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php (hunk 6 above)
Comment (fabianfrz) [reply src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: https://secure.php.net/manual/de/function.proc-open.php for the command calls and shellescapearg for the aguments.

File: src/opnsense/service/templates/OPNsense/WebGui/php.ini [hunk 7]
```diff
@@ -18,7 +18,7 @@ html_errors = Off
 zlib.output_compression = {% if system.webgui.compression|default('') != "" %}On{%else%}Off{%endif%}
 
 zlib.output_compression_level = {{system.webgui.compression|default('1')}}
-include_path = ".:/usr/local/etc/inc:/usr/local/www:/usr/local/opnsense/mvc:/usr/local/share/pear:/usr/local/opnsense/contrib"
+include_path = ".:/usr/local/share:/usr/local/etc/inc:/usr/local/www:/usr/local/opnsense/mvc:/usr/local/share/pear:/usr/local/opnsense/contrib"
```
Comment (fichtner) [inline src/opnsense/service/templates/OPNsense/WebGui/php.ini]: this would also go to master directly :)

Comment (machadovilaca) [reply src/opnsense/service/templates/OPNsense/WebGui/php.ini]: #2621

**Comment by fichtner** [comment]
At this point we need src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php to be a plugin PR and which dependencies you need so I can add them to the build process

I'm a bit worried about the lack of a configuration model... it just pushes a few values into the root node without prefix. Both Nextcloud and SCP use a model to make this less problematic for future growth.

File: src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php [hunk 8]
```diff
@@ -195,13 +195,11 @@ public function sendEmail($config, $confdata)
...
             file_put_contents($gpgPublicKeyFile, $gpgPublicKey);
+            proc_open('gpg --import < ' . $gpgPublicKeyFile);
         }
 
-
-        proc_open('gpg2 --import < ' . $gpgPublicKeyFile);
-
         proc_open('echo "' . $confdata . '"
-            | gpg2 --trust-model always --batch --yes --output backup --encrypt --recipient ' .
+            | gpg --trust-model always --batch --yes --output backup --encrypt --recipient ' .
```
Comment (fabianfrz) [inline src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: no, you should open the pipes and use them instead of the echo ... | part

Comment (machadovilaca) [reply src/opnsense/mvc/app/library/OPNsense/Backup/Mailer.php]: I think I fixed it now, thanks for the tips

**Comment by machadovilaca** [comment]
I will close this as I opened another PR in the plugins repository following https://github.com/opnsense/core/pull/2615#issuecomment-412059429 suggestion. https://github.com/opnsense/plugins/pull/788
